File: 20.Valid_Parentheses/main.go
package main

import "fmt"

type Solution struct{}

func (s Solution) IsValid(str string) bool {
    stack := []rune{}
    for _, char := range str {
        if char == '(' || char == '{' || char == '[' {
            stack = append(stack, char) // Push to stack
        } else {
            if len(stack) == 0 {
                return false
            }
            top := stack[len(stack)-1]
            stack = stack[:len(stack)-1] // Pop from stack
            if (char == ')' && top != '(') || 
               (char == '}' && top != '{') || 
               (char == ']' && top != '[') {
                return false
            }
        }
    }
    return len(stack) == 0
}

func main() {
    solution := Solution{}
    s := "({[]})"
    fmt.Println("Is Valid:", solution.IsValid(s)) // Output: true
}

File: 20.Valid_Parentheses/main.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

class Solution
{
public:
     bool isValid(string s)
     {
          stack<char> st;
          for (char c : s)
          {
               if (c == '(' || c == '{' || c == '[')
               {
                    st.push(c);
               }
               else
               {
                    if (st.empty())
                         return false;
                    char top = st.top();
                    st.pop();
                    if ((c == ')' && top != '(') ||
                        (c == '}' && top != '{') ||
                        (c == ']' && top != '['))
                    {
                         return false;
                    }
               }
          }
          return st.empty();
     }
};

int main()
{
     Solution solution;
     string s = "({[]})";
     cout << "Is Valid: " << (solution.isValid(s) ? "true" : "false") << endl;
     return 0;
}
